tests/drivers/clock_control/clock_control_api/src/test_clock_control.c

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2019, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ztest.h>
+#include <clock_control.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(test);
+
+#ifdef DT_INST_0_NORDIC_NRF_CLOCK_LABEL
+#include <drivers/clock_control/nrf_clock_control.h>
+#endif
+
+struct device_data {
+	const char *name;
+	u32_t startup_us;
+};
+
+static const struct device_data devices[] = {
+#ifdef DT_INST_0_NORDIC_NRF_CLOCK_LABEL
+	{
+		.name = DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_16M",
+		.startup_us = IS_ENABLED(CONFIG_SOC_SERIES_NRF91X) ? 3000 : 300
+	},
+
+	{
+		.name = DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_32K",
+		.startup_us = (CLOCK_CONTROL_NRF_K32SRC == NRF_CLOCK_LFCLK_RC) ?
+				1000 : 250000
+	}
+#endif
+};
+
+
+typedef void (*test_func_t)(const char *dev_name, u32_t startup_us);
+
+typedef bool (*test_capability_check_t)(const char *dev_name);
+
+static void setup_instance(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+	int err;
+
+	k_busy_wait(1000);
+	do {
+		err = clock_control_off(dev, 0);
+	} while (err == 0);
+	LOG_INF("setup done");
+}
+
+static void tear_down_instance(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+
+	clock_control_on(dev, NULL);
+}
+
+static void test_all_instances(test_func_t func,
+				test_capability_check_t capability_check)
+{
+	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
+		if ((capability_check == NULL) ||
+			capability_check(devices[i].name)) {
+			setup_instance(devices[i].name);
+			func(devices[i].name, devices[i].startup_us);
+			tear_down_instance(devices[i].name);
+			/* Allow logs to be printed. */
+			k_sleep(100);
+		}
+	}
+}
+
+/*
+ * Basic test for checking correctness of getting clock status.
+ */
+static void test_on_off_status_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int err;
+
+	zassert_true(dev != NULL, "%s: Unknown device", dev_name);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+
+	err = clock_control_on(dev, NULL);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true((status == CLOCK_CONTROL_STATUS_STARTING) ||
+			(status == CLOCK_CONTROL_STATUS_ON),
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	k_busy_wait(startup_us*1.5);
```

Comment:
Why you are using multiplier here? Aren't deadlines specified in the devices array?

Reply:
I want to take into account potential unalignment or imprecision of k_busy_wait.

Reply:
k_busy_wait is pretty accurate. IMHO it would be better to pass the values from the table directly.

Reply:
removed `1.5*` and trimmed startup times.